### PR TITLE
Tokenize row during in `training_step` 

### PR DESCRIPTION
**Pull request by qgallouedec**
# What does this PR do?

This PR is a first step towards enabling the use of any reward model (i.e. releasing the requirement to have the same tokenizer and chat template).

This change has no influence on results (screenshot 1) or learning speed (screenshot 2).
Screenshot 2024-09-25 at 11 11 41
Screenshot 2024-09-25 at 11 13 12

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you read the [contributor guideline](https://github.com/huggingface/trl/blob/main/CONTRIBUTING.md#create-a-pull-request),
      Pull Request section?
- [ ] Was this discussed/approved via a GitHub issue? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/trl/tree/main/docs).
- [ ] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

### CODE REVIEW

**Review by kashif — APPROVED** [review]

File: trl/trainer/online_dpo_trainer.py [hunk 1]
```diff
@@ -371,6 +352,13 @@ def get_eval_dataloader(self, eval_dataset: Optional[Union[str, Dataset]] = None
     def training_step(self, model: nn.Module, inputs: Dict[str, Union[torch.Tensor, Any]]) -> torch.Tensor:
         model.train()
 
+        # Apply chat template and tokenize the input
```
Comment (lewtun) [inline trl/trainer/online_dpo_trainer.py:355]: Would this make sense to include here?

```suggestion
        # Apply chat template and tokenize the input.
        # We do this on-the-fly to enable the use of reward models and policies with different tokenizers / chat templates.
```

**Review by lewtun — APPROVED** [review]
Thanks for the changes! Apart from a small comment, LGTM
